### PR TITLE
gh-109546: Add more tests for formatting floats and fractions

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-109546

### CODE REVIEW

File: Lib/test/test_fractions.py [hunk 1]
```diff
@@ -1220,6 +1220,30 @@ def test_invalid_formats(self):
                 with self.assertRaises(ValueError):
                     format(fraction, spec)
 
+    @requires_IEEE_754
+    def test_float_format_testfile(self):
+        from test.test_float import format_testfile
```
Comment (vstinner) [inline Lib/test/test_fractions.py]: I would prefer to just copy/paste the code to get the path here:

```py
format_testfile = os.path.join(test_dir, 'formatfloat_testcases.txt')
```

I prefer to not make tests inter-dependents. Sometimes, just importing a test has side effects.

File: Lib/test/test_fractions.py [hunk 2]
```diff
@@ -1220,6 +1220,30 @@ def test_invalid_formats(self):
...
+        with open(format_testfile, encoding="utf-8") as testfile:
+            for line in testfile:
+                if line.startswith('--'):
+                    continue
+                line = line.strip()
+                if not line:
+                    continue
```
Comment (vstinner) [inline Lib/test/test_fractions.py:1235]: I suppose that such logic is used in other files using formatfloat_testcases.txt. Is there a good place to share common code?

If we add mathdata/, can we add a helper function there which would iterate on the file and strip comments?

File: Lib/test/test_fractions.py [hunk 3]
```diff
@@ -1220,6 +1220,30 @@ def test_invalid_formats(self):
...
+
+                lhs, rhs = map(str.strip, line.split('->'))
+                fmt, arg = lhs.split()
+                if fmt == '%r':
+                    continue
+                with self.subTest(fmt=fmt, arg=arg):
+                    f = F(float(arg))
```
Comment (vstinner) [inline Lib/test/test_fractions.py:1243]: Maybe add a variable for ``fmt[1:]``.

**Review by vstinner — COMMENTED** [review]
I would prefer to not load test_float in test_fractions just to get a filename.

File: Lib/test/test_fractions.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_fractions.py:1235]: I was going to add it in `test_float`, but it is simple enough, so I just copied it.

File: Lib/test/test_fractions.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/test/test_fractions.py:1243]: Done.

**Comment by mdickinson** [comment]
+1 for the idea, and code changes LGTM.

**Review by mdickinson — APPROVED** [review]

**Comment by vstinner** [comment]
Oh. I tried to backport this change to 3.11 to prepare backporting PR #109512 to 3.11 and 3.12, but the fractions module doesn't support all formatting syntax in 3.11:

```
ERROR: test_float_format_testfile (test.test_fractions.FractionTest.test_float_format_testfile) (fmt='%.1g', arg='10')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vstinner/python/3.11/Lib/test/test_fractions.py", line 856, in test_float_format_testfile
    self.assertEqual(format(f, fmt2), rhs)
                     ^^^^^^^^^^^^^^^
TypeError: unsupported format string passed to Fraction.__format__
```

So this change cannot be backported.

**Comment by vstinner** [comment]
I partially backported the change to 3.11: only test_float changes.

**Pull request merged**